include/envoy/server/transport_socket_config.h

```diff
@@ -33,11 +33,31 @@ class TransportSocketConfigFactory {
 public:
   virtual ~TransportSocketConfigFactory() {}
 
+  /**
+   * @return ProtobufTypes::MessagePtr create empty config proto message. The transport socket
+   *         config, which arrives in an opaque google.protobuf.Struct message, will be converted
+   *         to JSON and then parsed into this empty proto.
+   */
+  virtual ProtobufTypes::MessagePtr createEmptyConfigProto() PURE;
+
+  /**
+   * @return std::string the identifying name for a particular TransportSocketFactoryPtr
+   *         implementation produced by the factory.
+   */
+  virtual std::string name() const PURE;
+};
+
+/**
+ * Implemented by each transport socket and registered via class RegisterFactory for upstream
```

Comment:
Do we have a requirement that each transport socket needs both ends or should this be "implemented by each transport socket which does upstream connections"?

GFE has some things it speaks upstream and not downstream.  With Envoy I have a hard time figuring out how we'd do testing without doing both so it's quite likely this comment should stand, maybe with an additional comment somewhere that it's required we always do both.

Reply:
No such requirement. Though I expect most implementations will have both upstream and downstream, nothing blocks an implementation with upstream only or downstream only, the interfaces and configuration support that. I will update the comments.

Tests could be done with external process, in some cases it could be done with existing other implementations, e.g. a transport socket is using other TLS implementation than BoringSSL.